### ISSUE TITLE
Why should we avoid using 'REQUIRED_CHILDREN' for a mutation?

### ISSUE THREAD

**Issue by ashinohara** [facebook/relay#236]
I am attempting use the relay-starter-kit to make a createWidgetMutation that looks like this:

```javascript
var createWidgetMutation = mutationWithClientMutationId({
  name: 'CreateWidget',
  inputFields: {
    name: {
        type: new GraphQLNonNull(GraphQLString)
    },
  },
  outputFields: {
    widget: {
      type: widgetType
    },
    widgetId: {
      type: GraphQLString
    },
  },
  mutateAndGetPayload: (payload) => {
    console.log('createWidget', payload);
    var result = {widget: {name: 'test widget'}, widgetId: 1};
    return Promise.resolve(result);
  }
});
```

When I use FIELDS_CHANGE as the mutation config, the widget object does not get passed back in the payload, but widgetId is passed back.  When I use REQUIRED_CHILDREN, the object does get passed back, but the documentation for REQUIRED_CHILDREN says to avoid using this type.  What should I use instead of REQUIRED_CHILDREN?

My mutation looks like this for REQUIRED_CHILDREN:

```javascript
export default class CreateWidgetMutation extends Relay.Mutation {
    static fragments = {
        widget: () => Relay.QL`
            fragment on Widget {
                name
            }
        `,
    };

    getMutation() {
        return Relay.QL`
        mutation {
            createWidget
        }`;
    }
    getVariables() {
        return {
            name: this.props.name
        };
    }
    getFatQuery() {
        return Relay.QL`
        fragment on CreateWidgetPayload {
            widget,
            widgetId
        }
        `;
    }
    getConfigs() {
        return [{
            type: 'REQUIRED_CHILDREN',
            children: [Relay.QL`
            fragment on CreateWidgetPayload {
                widget {
                    name
                },
                widgetId
            }
            `]
        }];
    }
}
```

and the getConfigs looks like this for FIELDS_CHANGE:

```javascript
    getConfigs() {
        return [{
            type: 'FIELDS_CHANGE',
            fieldIDs: {
                widget: null,
                widgetId: null
            },
        }];
    }
```

**Comment by yungsters** [comment]
Thanks for bringing this up. Indeed, you do need to use `CHILDREN_REQUIRED` because `widget` does not have a parent field (from my understanding) that the view could possibly have already fetched or rendered.

I spoke with @yuzhi and we came to the conclusion that:

 - `CHILDREN_REQUIRED` should be renamed to something like `EXTRA_FRAGMENT`.
 - We need to revise the code documentation (and add website documentation) for this and its valid use cases.

I hope this answers your question. I'll open a separate issue to rename and revise documentation.

----

Separately, I noticed you have `widget` and `widgetId`. Why not just fetch `widget { id }`?

**Comment by yungsters** [comment]
Filed #237.

**Comment by mhart** [comment]
> Separately, I noticed you have widget and widgetId. Why not just fetch widget { id }?

@yungsters my guess is because global IDs are supposed to be opaque :smile_cat: Is there a current recommended method to retrieve the non-global ID of an entity/node other than adding an extra field?

**Comment by yungsters** [comment]
@mhart No, but `widget { id }` should be the opaque global ID. If for whatever reason, you actually need the non-global ID, I would actually recommend making *that* available via an extra field. Does that make sense?

**Comment by mhart** [comment]
That's what I'm saying.

**Comment by mhart** [comment]
(ie, my guess is that @ashinohara added that field to get access to the ID of the node – because there's no other way to do it!)

**Comment by mhart** [comment]
I think it's just one aspect of Relay that's confusing – that the IDs on your objects will actually be transformed and no longer available to the client... unless you add an extra field.

**Comment by yungsters** [comment]
Oooh, I see what you mean. Sorry for the confusion.

Anyway, I think the questions here are resolved? We can follow-up on #237.

**Comment by ashinohara** [comment]
@yungsters Yes, we can consider this resolved.  The reason I was retrieving widgetId to begin with was to debug the behavior of REQUIRED_CHILDREN vs. FIELDS_CHANGE.
